### DISCUSSION TITLE
Verify smartcontract on sepolia-era.zksync.network with hardhat plugin

### DISCUSSION THREAD

**Discussion by cytadela8** [zkSync-Community-Hub/zksync-developers#732] [Dev Tooling] [answered]
### Environment

Testnet

### Select the Dev Tool you are using

Hardhat Plugins

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

How do I verify a smartcontract on https://sepolia-era.zksync.network/? On https://sepolia.explorer.zksync.io/ I was able to verify using hardhat-zksync-verify without any problems. Is there an easy way to do this on https://sepolia-era.zksync.network/?

Looking at the docs for explorer it seems similar to Etherscan and I would need an API key, but [the link here is broken](https://docs.zksync.network/zksync-testnet/api-endpoints/contracts#verify-source-code:~:text=API%20PRO%20tier.-,Requires%20a%20valid%20zkSync%20Era%20Sepolia%20Explorer%20API%20key%2C%20it%20will%20be%20rejected%20otherwise,-Only%20supports%20HTTP)

**Comment by alhenaai** [comment]
To verify a smart contract on the Sepolia Era network of ZKsync using the Hardhat plugin, you need to ensure that your `hardhat.config.ts` file is correctly configured for the Sepolia Era network. Here are the steps you can follow:

1. **Configure the Network in `hardhat.config.ts`**:
   Ensure that your network configuration includes the `zksync` flag set to true and the correct `verifyURL`. Here's an example configuration for the Sepolia Era network:

   ```typescript
   zkSyncSepoliaTestnet: {
     url: "https://sepolia.era.zksync.dev",
     ethNetwork: "sepolia",
     zksync: true,
     verifyURL: "https://explorer.sepolia.era.zksync.dev/contract_verification",
   },
   ```

2. **Use the Hardhat Command**:
   To verify your contract, use the Hardhat command with the contract's address. For example:
   ```bash
   npx hardhat verify --network zkSyncSepoliaTestnet YOUR_CONTRACT_ADDRESS
   ```

3. **API Key Requirement**:
   If the Sepolia Era explorer requires an API key similar to Etherscan, you will need to obtain this from the ZKsync Sepolia Explorer's documentation or support. Ensure that your `verifyURL` is correctly pointing to the API endpoint that supports contract verification.

If you encounter issues with broken links in the documentation or need an API key and cannot find information, I recommend reaching out directly to ZKsync support or checking the updated documentation at [ZKsync Docs](https://docs.zksync.io/).

For further details on using the Hardhat plugin for ZKsync, you can refer to the [Hardhat ZKsync Verify Plugin Documentation](https://docs.zksync.io/build/tooling/hardhat/plugins/hardhat-zksync-verify-vyper).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by kaymomin** [reply, marked as answer]
Both [ZKsync Explorer](https://sepolia.explorer.zksync.io/) and [Etherscan explorer for Era](https://sepolia-era.zksync.network/) have different APIs, which makes the verification process a bit different for each.

The `hardhat-zksync-verify` plugin was originally built to verify contracts specifically on the ZKsync Explorer since, at the time, we didn’t have an Etherscan explorer for Era. However, we're actively working on updating the plugin to support both explorers. 

You can expect the new version to be released later this week or early next week.

If you need a solution right now, the latest foundry-zksync release already supports both explorers. Just use `forge verify-contract` to verify your contracts.

**Reply by alhenaai** [reply]
Thank you for the update on the `hardhat-zksync-verify` plugin and the information about the Foundry tool's capabilities. For immediate verification needs on both the ZKsync Explorer and the Etherscan explorer for Era, using Foundry's `forge verify-contract` command is a great alternative.

Here’s how you can use Foundry to verify your contracts:

1. **Install Foundry**:
   If you haven't already installed Foundry, you can do so by following the instructions [here](https://book.getfoundry.sh/getting-started/installation.html).

2. **Verify Your Contract**:
   Use the `forge verify-contract` command with the appropriate parameters. Here's an example command:
   ```bash
   forge verify-contract --chain-id YOUR_CHAIN_ID --compiler-version YOUR_COMPILER_VERSION --constructor-args ARGS_FILE_PATH YOUR_CONTRACT_ADDRESS "YourContract.sol:YourContract" YOUR_ETHERSCAN_API_KEY
   ```

   Replace `YOUR_CHAIN_ID`, `YOUR_COMPILER_VERSION`, `ARGS_FILE_PATH`, `YOUR_CONTRACT_ADDRESS`, and `YOUR_ETHERSCAN_API_KEY` with the relevant details for your contract and network.

3. **Wait for Plugin Update**:
   Keep an eye on the official ZKsync documentation or announcements for the updated `hardhat-zksync-verify` plugin that will support both explorers.

This approach should help you manage your immediate needs for contract verification while waiting for the updated Hardhat plugin. If you encounter any issues or have further questions, feel free to reach out for more assistance.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.
